rfcs/0100-sign-commits.md

```diff
@@ -0,0 +1,345 @@
+- https://github.com/guix-mirror/guix/commits/master
+- https://guix.gnu.org/manual/en/html_node/Commit-Access.html
+- https://guix.gnu.org/manual/en/html_node/Channel-Authentication.html
+
+It is also done by most conventional distros, however, they sign
+the output from the compilation rather than the source code, since
+they distribute the output from the compilation.
+
+This change would also allow using 3rd-party mirrors of Nixpkgs,
+since we can verify the source code independent of where it came from.
+
+This RFC was motivated by work during Summer of Nix, specifically,
+the Libre SoC maintainer requested that source integrity be
+maintained.
+
+# Detailed design
+[design]: #detailed-design
+
+## Signing the commits
+[signing]: #signing-the-commits
+
+The following will be required from 2022-03-01T00:00Z on,
+but is strongly recommended from now on.
+
+Currently merging is done through GitHub, which would not work in this case.
+Merging would have to happen outside of GitHub, so that you can sign the commits,
+since currently the API doesn't support signing commits with your own PGP
+key: https://github.com/cli/cli/issues/1318.
+
+We would make a script for merging pull requests instead, something like:
+```bash
+#!/bin/sh
+# nixpkgsmerge
+
+git switch master
+git pull
+gh pr checkout $1
+git switch master
+git merge -m "Merged #$1" --no-ff --gpg-sign -
+git push origin master
+gh pr close $1
+```
+Then
+```bash
+nixpkgsmerge 123
+```
+
+For mobile platforms, you would use [nix-on-droid](https://github.com/t184256/nix-on-droid)
+to run the script instead.
+
+A merge commit must always be made so that there is something to sign.
+
+A git hook like the one presented here is also recommended,
+so that unsigned commits aren't pushed by accident.
+This would be put in `/.git/hooks/pre-push` of the repository to
+be committed to.
+```bash
+#!/bin/sh
+
+# Inspiration taken from https://github.com/guix-mirror/guix/blob/master/etc/git/pre-push.
+# Put this in .git/hooks/pre-push.
+set -xe
+
+# This is the "empty hash" used by Git when pushing a branch deletion.
+z40=0000000000000000000000000000000000000000
+
+while read local_ref local_hash remote_ref remote_hash
+do
+  # Branch deletion; no signature to verify.
+  if [ "$local_hash" != $z40 ]
+  then
+    # Check repositories under github.com/NixOS.
+    case "$2" in
+        *github.com*NixOS/*)
+      # TODO: More thorough check.
+      git verify-commit HEAD
+      ;;
+    esac
+  fi
+done
+```
+
+The digest algorithm for GPG (if you're using GPG) must be set to something that isn't SHA-1,
+as [it isn't secure anymore](https://eprint.iacr.org/2020/014.pdf), for example by putting the following in
+`$HOME/.gnupg/gpg.conf`:
+```
+digest-algo sha512
+```
+
+Committers will need to put their PGP keys in
+https://github.com/NixOS/nixpkgs/blob/master/maintainers/maintainer-list.nix,
+and add their keys to public key servers.
+
+The following might not be implemented before 2022-03-01T00:00Z,
```

Comment:
It is not significant. I chose it arbitrarily after considering how much time we might need to get proper tooling for signing merges. It could be anything else.